### PR TITLE
✨ Add SideNav component

### PR DESCRIPTION
**Pull request by MaxMuecket**
- Add SideNav.tsx
- Add SideNav.stories.tsx
- Add SideNav.module.css

https://pocket-garde-side-menu-ps7xats.herokuapp.com/storybook/?path=/story/component-sidenav--garden

### CODE REVIEW

**Review by J-Bannach — APPROVED** [review]
good job! 🏅

File: src/app/components/SideNav/SideNav.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+import styles from './SideNav.module.css';
+
+export type SideNavProps = {
+  children: string;
+  garden?: string;
+  newPlant?: string;
+  plantList?: string;
+  newTask?: string;
+  gardening?: string;
```
Comment (lmachens) [inline src/app/components/SideNav/SideNav.tsx:12]: Do you have use-cases where you give these properties different values?
It looks like a fix navigation.

A more common way would be to add a property like `navigation`:
`navigation: React.ReactNode;`

with 
```tsx
<div className={styles.sideNav__menu_items}>
          <button
            className={styles.sideNav__button_active}
            onClick={showSideNav}
          >
            {children}
          </button>
          {navigation}
        </div>
```

and 
```tsx
<SideNav navigation={
  <>
    <Link to="/garden">{garden}</Link>
    <Link to="/new-plant">{newPlant}</Link>
    <Link to="/plant-list">{plantList}</Link>
    <Link to="/new-task">{newTask}</Link>
    <Link to="/gardening">{gardening}</Link>
  </>
} />
```

See https://reactwithhooks.netlify.app/docs/composition-vs-inheritance.html

Another option is to have a fix navigation without any property.

File: src/app/components/SideNav/SideNav.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import React, { useState } from 'react';
+import { Link } from 'react-router-dom';
+
+import styles from './SideNav.module.css';
+
+export type SideNavProps = {
+  children: string;
```
Comment (lmachens) [inline src/app/components/SideNav/SideNav.tsx:7]: Usually, `children` is not only a `string`.
You set the content of the button, which shows the menu.
Maybe call it `showSideNavLabel: string;`?

If this is always the same, you don't need a property for this.

File: src/app/components/SideNav/SideNav.module.css [hunk 3]
```diff
@@ -0,0 +1,78 @@
+body {
+  background-color: var(--background-color);
+}
```
Comment (lmachens) [inline src/app/components/SideNav/SideNav.module.css:3]: You like to set global CSS in a CSS module. This effect other components too and should be in your globals.css.

**Review by lmachens — CHANGES_REQUESTED** [review]
good job!

File: src/app/components/SideNav/SideNav.tsx (hunk 1 above)
Comment (MaxMuecket) [reply src/app/components/SideNav/SideNav.tsx:12]: I wanted to build a menu that includes all links to the other sides without the link from the active side.
That's the reason why I use the properties to decide which link should be on the page.

Comment (lmachens) [reply src/app/components/SideNav/SideNav.tsx:12]: This sounds uncommon, do you know websites/apps doing the same thing?
It's more common to highlight the active page.

Anyway, if you like to do that, you could make use of the [`useLocation`](https://reactrouter.com/web/api/Hooks/uselocation) hook to identify the current path and hide the link you don't want to display. 

```tsx
const location = useLocation();

return (
  <>
    {location.pathname !== '/' && <Link to="/">Home</Link>}
    {location.pathname !== '/plants' && <Link to="/plants">Plants</Link>}
  </>
);

Comment (MaxMuecket) [reply src/app/components/SideNav/SideNav.tsx:12]: That sounds nice. Thanks for your help. 

For this example, I have to split the component in two, is that right?

with
```
<div className={styles.sideNav__menu_items}>
          <button
            className={styles.sideNav__button_active}
            onClick={showSideNav}
          >
            {children}
          </button>
          {navigation}
        </div>
```
and

```
<SideNav navigation={
  <>
    <Link to="/garden">{garden}</Link>
    <Link to="/new-plant">{newPlant}</Link>
    <Link to="/plant-list">{plantList}</Link>
    <Link to="/new-task">{newTask}</Link>
    <Link to="/gardening">{gardening}</Link>
  </>
} />
```

Comment (lmachens) [reply src/app/components/SideNav/SideNav.tsx:12]: Not necessary, you can still add a static navigation to your `SideNav`

Comment (MaxMuecket) [reply src/app/components/SideNav/SideNav.tsx:12]: You don't mean that, do you? 

```
import React, { useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

import styles from './SideNav.module.css';

export type SideNavProps = {
  showSideNavLabel?: string;
  navigation: React.ReactNode;
};

function SideNav({ showSideNavLabel, navigation }: SideNavProps): JSX.Element {
  const [active, setActive] = useState(false);

  const showSideNav = () => setActive(!active);

  const location = useLocation();

  return (
    <>
      <div className={styles.sideNav}>
        <button className={styles.sideNav__button} onClick={showSideNav}>
          {showSideNavLabel}
        </button>
      </div>
      <nav
        className={
          active ? `${styles.sideNav__menu_active}` : `${styles.sideNav__menu}`
        }
      >
        <div className={styles.sideNav__menu_items}>
          <button
            className={styles.sideNav__button_active}
            onClick={showSideNav}
          >
            {showSideNavLabel}
          </button>
          {navigation}
        </div>
        <SideNav
          navigation={
            <>
              {location.pathname !== '/garden' && (
                <Link to="/garden">Garden</Link>
              )}
              {location.pathname !== '/new-plant' && (
                <Link to="/new-plant">NewPlant</Link>
              )}
              {location.pathname !== '/plant-list' && (
                <Link to="/plant-list">PlantList</Link>
              )}
              {location.pathname !== '/new-task' && (
                <Link to="/new-task">NewTask</Link>
              )}
              {location.pathname !== '/gardening' && (
                <Link to="/gardening">Gardening</Link>
              )}
            </>
          }
        />
      </nav>
    </>
  );
}

export default SideNav;
```

Comment (lmachens) [reply src/app/components/SideNav/SideNav.tsx:12]: Now you have a `SideNav` inside a `SideNav`. I thought about this:

```tsx
import React, { useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

import styles from './SideNav.module.css';

export type SideNavProps = {
  showSideNavLabel?: string;
  navigation: React.ReactNode;
};

function SideNav({ showSideNavLabel, navigation }: SideNavProps): JSX.Element {
  const [active, setActive] = useState(false);

  const showSideNav = () => setActive(!active);

  const location = useLocation();

  return (
    <>
      <div className={styles.sideNav}>
        <button className={styles.sideNav__button} onClick={showSideNav}>
          {showSideNavLabel}
        </button>
      </div>
      <nav
        className={
          active ? `${styles.sideNav__menu_active}` : `${styles.sideNav__menu}`
        }
      >
        <div className={styles.sideNav__menu_items}>
          <button
            className={styles.sideNav__button_active}
            onClick={showSideNav}
          >
            {showSideNavLabel}
          </button>
          {navigation}
        </div>
          {location.pathname !== '/garden' && (
            <Link to="/garden">Garden</Link>
          )}
          {location.pathname !== '/new-plant' && (
            <Link to="/new-plant">NewPlant</Link>
          )}
          {location.pathname !== '/plant-list' && (
            <Link to="/plant-list">PlantList</Link>
          )}
          {location.pathname !== '/new-task' && (
            <Link to="/new-task">NewTask</Link>
          )}
          {location.pathname !== '/gardening' && (
            <Link to="/gardening">Gardening</Link>
          )}
      </nav>
    </>
  );
}

export default SideNav;
```

Now, you can think about doing further optimizations like a mapped list of links (like in your `App.tsx` with the `routes`) to reduce similar code.

Comment (MaxMuecket) [reply src/app/components/SideNav/SideNav.tsx:12]: I'm not sure how to implement the `useLocation()` hook at the mapped list of links.
Do you have a tip for me, please?

```
import path from 'path';
import React, { useState } from 'react';
import { Link, LinkProps, useLocation } from 'react-router-dom';
import GardenPage from '../../pages/GardenPage/GardenPage';

import styles from './SideNav.module.css';

export type SideNavProps = {
  showSideNavLabel: string;
};

type CustomLinkProps = LinkProps & {
  linkName: string;
};

const links: CustomLinkProps[] = [
  { to: '/garden', linkName: 'Garden' },
  { to: '/new-plant', linkName: 'NewPlant' },
  { to: '/plant-list', linkName: 'PlantList' },
  { to: '/new-task', linkName: 'NewTask' },
  { to: '/gardening', linkName: 'Gardening' },
];

function SideNav({ showSideNavLabel }: SideNavProps): JSX.Element {
  const [active, setActive] = useState(false);

  const showSideNav = () => setActive(!active);

  const location = useLocation();

  return (
    <>
      <div>
        <button className={styles.sideNav__button} onClick={showSideNav}>
          {showSideNavLabel}
        </button>
      </div>
      <nav
        className={
          active ? `${styles.sideNav__menu_active}` : `${styles.sideNav__menu}`
        }
      >
        <div className={styles.sideNav__menu_items}>
          <button
            className={styles.sideNav__button_active}
            onClick={showSideNav}
          >
            {showSideNavLabel}
          </button>
          {links.map(({ ...linkProps }) => (
            <Link key={linkProps.linkName} {...linkProps}>
              {linkProps.linkName}
            </Link>
          ))}
        </div>
      </nav>
    </>
  );
}

export default SideNav;
```

Comment (lmachens) [reply src/app/components/SideNav/SideNav.tsx:12]: You can filter the `links` first to have only the links you want to display:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter

```tsx
links.filter(link => link.to !== location.pathname)
```

Comment (MaxMuecket) [reply src/app/components/SideNav/SideNav.tsx]: Thanks a lot! ✌️

**Pull request merged**